### PR TITLE
fix: added support to border and radius for canvas2d

### PR DESCRIPTION
**Pull request by pecoram**

### CODE REVIEW

File: src/core/renderers/canvas/CanvasCoreRenderer.ts [hunk 1]
```diff
@@ -134,8 +135,22 @@ export class CanvasCoreRenderer extends CoreRenderer {
     const hasClipping = clippingRect.width !== 0 && clippingRect.height !== 0;
     const hasGradient = colorTl !== colorTr || colorTl !== colorBr;
     const radius = quad.shader ? getRadius(quad) : 0;
+    const border = quad.shader ? getBorder(quad) : undefined;
+    const borderTop = quad.shader ? getBorder(quad, 'Top') : undefined;
```
[the remark this answers is not in the document]
Comment (pecoram) [reply src/core/renderers/canvas/CanvasCoreRenderer.ts:139]: I am undecided whether it is better to give freedom to override one of the borders (eg. border+borderBottom) or make them exclusive

**Comment by pecoram** [comment]
there is a problem with the position of the border... on the webgl renderer the border is inside (which I think is wrong or at most to be made configurable) while on the canvas renderer it is centered

File: src/core/renderers/canvas/CanvasCoreRenderer.ts [hunk 2]
```diff
@@ -134,8 +135,26 @@ export class CanvasCoreRenderer extends CoreRenderer {
     const hasClipping = clippingRect.width !== 0 && clippingRect.height !== 0;
     const hasGradient = colorTl !== colorTr || colorTl !== colorBr;
     const radius = quad.shader ? getRadius(quad) : 0;
+    const border = quad.shader ? getBorder(quad) : undefined;
+    const borderTop =
+      !border?.width && quad.shader ? getBorder(quad, 'Top') : undefined;
```
Comment (elsassph) [inline src/core/renderers/canvas/CanvasCoreRenderer.ts:140]: You could avoid this complex test (esp when transpiled to ES5) by having `getBorder` returning `undefined` when there is no width.

Comment (elsassph) [reply src/core/renderers/canvas/CanvasCoreRenderer.ts:140]: Also maybe:
```typescript
    let tryBorderShader = Boolean(quad.shader);
    const border = tryBorderShader ? getBorder(quad) : undefined; // assuming `undefined` if no `border.width`
    if (boder) tryBorderShader = false;
    const borderTop = tryBorderShader ? getBorder(quad, 'Top') : undefined;
    //etc
```

File: src/core/renderers/canvas/CanvasCoreRenderer.ts [hunk 3]
```diff
@@ -134,8 +135,26 @@ export class CanvasCoreRenderer extends CoreRenderer {
...
+      !border?.width && quad.shader ? getBorder(quad, 'Bottom') : undefined;
+    const borderLeft =
+      !border?.width && quad.shader ? getBorder(quad, 'Left') : undefined;
 
-    if (hasTransform || hasClipping || radius) {
+    if (
+      hasTransform ||
+      hasClipping ||
+      radius ||
+      border ||
```
Comment (elsassph) [inline src/core/renderers/canvas/CanvasCoreRenderer.ts:152]: Do we need to save the context for borders?

**Comment by wouterlucas** [comment]
Nice, thanks for the contribution! I like how you've isolated it in the Canvas implementation and thanks for the extended review @elsassph!

Heads up @jfboeve has been working on a redesign on the shaders/effects portion of the renderer. Stepping away from the Dynamic Shader and having tree shakeable shaders.

One of the topics we discussed is shaders and Canvas2D and we're keeping those separate. Limiting Effects to WebGL1/WebGL2 only as it does do not make sense to combine. Trying to come up with something to caters to both WebGL & Canvas2D would only lead to overly complicated effects definition for only a small number of applicable effects for Canvas2D anyway.

We'll have to refactor this PR a little bit to make it fit with the new effects/shaders structure down the road.

**Review by wouterlucas — APPROVED** [review]

File: src/core/renderers/canvas/CanvasCoreRenderer.ts [hunk 4]
```diff
@@ -211,6 +214,94 @@ export class CanvasCoreRenderer extends CoreRenderer {
...
+      } else {
+        ctx.strokeRect(
+          tx + borderInnerWidth,
+          ty + borderInnerWidth,
+          width - borderWidth,
+          height - borderWidth,
+        );
+      }
+      ctx.globalAlpha = 1;
+    } else {
```
Comment (elsassph) [inline src/core/renderers/canvas/CanvasCoreRenderer.ts:244]: `else if (hasQuadShader) {` at least

**Review by elsassph — APPROVED** [review]

File: src/core/renderers/canvas/shaders/UnsupportedShader.ts [hunk 5]
```diff
@@ -27,9 +27,9 @@ export class UnsupportedShader extends CoreShader {
   constructor(shType: string) {
     super();
     this.shType = shType;
-    if (shType !== ROUNDED_RECTANGLE_SHADER_TYPE) {
-      console.warn('Unsupported shader:', shType);
-    }
+    // if (shType !== ROUNDED_RECTANGLE_SHADER_TYPE) {
```
Comment (elsassph) [inline src/core/renderers/canvas/shaders/UnsupportedShader.ts:30]: Should that be removed?

Comment (pecoram) [reply src/core/renderers/canvas/shaders/UnsupportedShader.ts:30]: let's decide, having added the missing “shaders” I think it is no longer necessary. In addition, this warning makes the inspector unusable

**Pull request merged**